### ISSUE TITLE
proposal: support iptc keywords

### ISSUE THREAD

**Issue by smolin** [xemle/home-gallery#11]
Hi, this project is awesome, thank you very much for doing it!

For me the thing that would make it perfect is for IPTC keywords to be visible and searchable.

I might be able to help with development in a small way. I've successfully built a new docker image with a small change to verify that the build is working.

**Comment by xemle** [comment]
> Hi, this project is awesome, thank you very much for doing it!

Thank you very much.

For curiosity: How did you find this project and what is your use case you are trying to solve with this gallery?

> For me the thing that would make it perfect is for IPTC keywords to be visible and searchable.

That should be fairly easy to improve.

> I might be able to help with development in a small way. I've successfully built a new docker image with a small change to verify that the build is working.

Since you offered some help I will give you some points to start. I do not know your level of coding/reverse engineering. If you need more information, please ask questions ;-) My assumption of prerequisite is a rough understanding of the [internal architecture](https://docs.home-gallery.org/internals/index.html) and js...

The IPTC tags are already extracted via exiftool and are stored (as general exiftoool output) in the storage per image as `...-exif.json` file. The missing part will be the add them in the database build step to use them and to add them to the search term.

- inspect your `...-exif.json` file which data you need to extract. The tool [jq](https://stedolan.github.io/jq/) helps a lot here
- the database build step can be found [here](https://github.com/xemle/home-gallery/blob/master/packages/database/src/map-media.js#L114).
- the searchable text is build [here](https://github.com/xemle/home-gallery/blob/master/packages/query/src/index.js#L8)

To be sure that you are on the right track, I propose to work with VS code on the source directly using `npm run build`, maybe only `npm run build -- --scope '*/{database,query}' to  limit the build updates on database and query package only. Setup a small gallery with IPTC tags, check that the database is updated correctly and test the query with plain CLI call for the export, which supports the query, too via `./gallery.js export ... -q <your iptc keyword>`. If that is working, the webapp can be build and used.

Do not hesitate to ask any question and have fun.

**Comment by smolin** [comment]
Thank you for the clear feedback! I do hope to give this a try, though I don't have a lot of time for personal projects. If you get to it before me, I won't be upset :-) It's the sort of thing I do all day for work, so often in the evening I don't want to look at the computer any more ...

I too found your project through awesome-selfhosted, after inspecting some of the options suggested by https://alternativeto.net/software/google-photos/ --- I have been using Amazon Photos (I have Prime) but the lack of open programatic access is a show-stopper for me (I'm not about to waste my time with their proprietary sync software).

I'm hoping to use home-gallery for visual sorting and finding images, and backup to Backblaze B2 via rclone.

**Comment by xemle** [comment]
Thank you for your response and your details.

Regarding IPTC: Would you mind your IPTC tags which you are using? Either by exiftool command args or via an example file containing these data. Than this topic has the change to get implemented if time and mood is available.

Regarding your use case finding alternatives to google photos: Which other gallery software have you tried/will you try? Are you aware of [awesome-selfhosted photo galleries](https://github.com/awesome-selfhosted/awesome-selfhosted#photo-and-video-galleries)? What are the arguments for HomeGallery as web photo gallery?

Side note and regarding backup: rclone sounds like syncing private photos directly to the cloud? If so I can recommend to use an encrypted backup tool like [borg backup](borgbackup.readthedocs.io/) or [restic](https://restic.net). I use the first since years while the later supports better cloud storage provider like Backblaze or [wasabi](https://wasabi.com).

**Comment by smolin** [comment]
Thanks, I'll definitely give borg and restic another look.

awesome-selfhosted photo galleries is in fact how I found you! I've considered or tried ImageStore, LibrePhotos, Lychee, Photo Stream, PhotoPrism and Photoview, and possibly others (and non-self-hosted Amazon, Google, Flickr, Smugmug, Dropbox, Pbase, Microsoft whatever ... probably others I'm not remembering). The non-self-hosted options all dropped out because of either price or non-support for rclone (restic, etc). The self-hosted ones either failed to run for me right away (usually using a Docker method) or looked too complicated to manage (separate MySql server, for example).

home-gallery ticks a lot of boxes: all-pictures-by-date view; optional metadata view; similar image view; metadata search; filename display; object recognition; clean simple aesthetic; easy setup; works with filesystem as-is.

As for keywords, sadly there doesn't seem to be much of a consensus, but I don't have elaborate plans (no hierarchies of keywords, for example). When I add keywords "thing1" and "thing2" via Digikam and then extract them with exiftool, I get

```
exiftool -s -G foo.jpg |  grep thing1
[XMP]           Categories                      : <Categories><Category Assigned="1">thing1</Category><Category Assigned="1">thing2</Category></Categories>
[XMP]           LastKeywordXMP                  : thing1, thing2
[XMP]           TagsList                        : thing1, thing2
[XMP]           HierarchicalSubject             : thing1, thing2
[XMP]           CatalogSets                     : thing1, thing2
[XMP]           Subject                         : thing1, thing2
[IPTC]          Keywords                        : thing1, thing2
```

The amazingly in-depth article https://www.carlseibert.com/xmp-iptciim-or-exif-which-is-preferred/ suggests (in the comments) 'For a reading order, I usually recommend a simple XMP > IIM > Exif. [...] “Subject” is the XMP place for keywords.'

Thinking out loud here, I'd probably go with XMP Subject as a comma-separated list of keywords; if it's blank, try IPTC Keywords; if that's blank, then bail.

BTW if all the pics in the demo gallery are yours, I'm jealous of how much you've traveled.

**Comment by xemle** [comment]
Thank you so much for the flowers regarding this gallery and comparison to other solutions. That is very good to hear. Since I am developing the gallery I do not compare other solutions in detail - only reading their feature list and clicking through their demos.

Further, I do not prefer social media and marketing. According your description and the current stealth mode of gallery it could need some blogging articles, tweet or comments on reddit for publicity, etc. Do you have any possibilities or relations to provide some public infos/feedback?


Regarding IPTC meta data: The main file format for the gallery is JSON. exiftool is also called with `-json` argument to provide this format (via [exiftool-vendored package](https://www.npmjs.com/package/exiftool-vendored). Could you create a gist with your sample JSON meta data output from your JPG and XMP files? Or send your files in a private mail. Than it would be the easiest way to discuss these properties and their prioritizations.

**Comment by smolin** [comment]
I'm not much into the social media scene either, but I'll give it some thought. 

In the meantime, here's the output from `exiftool -s -G -json Downloads/rgbNumbers.jpg`
```
[{
  "SourceFile": "Downloads/rgbNumbers.jpg",
  "ExifTool:ExifToolVersion": 12.00,
  "ExifTool:Warning": "[minor] Fixed incorrect URI for xmlns:MicrosoftPhoto",
  "File:FileName": "rgbNumbers.jpg",
  "File:Directory": "Downloads",
  "File:FileSize": "8.6 kB",
  "File:FileModifyDate": "2021:08:24 08:46:02-07:00",
  "File:FileAccessDate": "2021:08:24 08:46:04-07:00",
  "File:FileInodeChangeDate": "2021:08:24 08:46:02-07:00",
  "File:FilePermissions": "rwxr-xr-x",
  "File:FileType": "JPEG",
  "File:FileTypeExtension": "jpg",
  "File:MIMEType": "image/jpeg",
  "File:CurrentIPTCDigest": "adbed99074e84af386f5b906de926d94",
  "File:Comment": "Created with The GIMP",
  "File:ImageWidth": 256,
  "File:ImageHeight": 256,
  "File:EncodingProcess": "Baseline DCT, Huffman coding",
  "File:BitsPerSample": 8,
  "File:ColorComponents": 3,
  "File:YCbCrSubSampling": "YCbCr4:2:0 (2 2)",
  "JFIF:JFIFVersion": 1.01,
  "JFIF:ResolutionUnit": "inches",
  "JFIF:XResolution": 72,
  "JFIF:YResolution": 72,
  "XMP:XMPToolkit": "XMP Core 4.4.0-Exiv2",
  "XMP:Categories": "<Categories><Category Assigned=\"1\">thing1</Category><Category Assigned=\"1\">thing2</Category></Categories>",
  "XMP:TagsList": ["thing1","thing2"],
  "XMP:LastKeywordXMP": ["thing1","thing2"],
  "XMP:HierarchicalSubject": ["thing1","thing2"],
  "XMP:CatalogSets": ["thing1","thing2"],
  "XMP:Subject": ["thing1","thing2"],
  "IPTC:CodedCharacterSet": "UTF8",
  "IPTC:Keywords": ["thing1","thing2"],
  "Composite:ImageSize": "256x256",
  "Composite:Megapixels": 0.066
}]
```
And the jpeg file
![rgbNumbers](https://user-images.githubusercontent.com/4467247/130648196-df4c8729-647b-41b1-b7bb-11cbe8adf340.jpg)

**Comment by xemle** [comment]
Thanks for your details.

Which tags are of interest for you? Would you mind to provide a mapping to the target media object of HomeGallery?

If you think a direct chat or video call would be helpful to clarify these things quickly, let me know...

> BTW if all the pics in the demo gallery are yours, I'm jealous of how much you've traveled.

Ahm, yes. 98% of them I have pressed the shutter by myself :-) I did one word travel and a longer trip through Europe

**Comment by xemle** [comment]
Thank you for your input. `IPTC:Keywords": ["thing1","thing2"]` should work and is enough information I need.

I could also need this feature since I have tagged my older images via these IPTC tags ;-)

Regarding a video call: I am in UTC+2 currently. Would you have time tomorrow morning, August 29th between 11 am an 1 pm (your time)? That would be 8 pm and 10 pm my time. We could meet us in the [community chat](https://gitter.im/home-gallery/community). Or provide another time frame...

**Comment by xemle** [comment]
Just want to give a quick update: I am currently working on this issue to support IPTC keywords.

Keywords from a main file was an easy catch. However I would like to support sidecar files like xmp files, too. If the sidecars stay static and are not changing it is also not problematic. But if the sidecar changes it becomes tricky at the moment and I was not able yet to find a good solution for it.

If done well I hope I could also catch the sequence feature with it where a series of similar photos are grouped to one (e.g by an unsorted multiple shot try on mobile phones or time laps pictures).

Stay tuned.

**Comment by xemle** [comment]
Hi. I've published my feature branch `feature/exif-tags` of IPTC keyword support. You can download binaries from https://dl.home-gallery.org/dist/unstable/latest of it.

It should support xmp sidecar files and also detect and import changes. To test, please run a full import to extract xmp files and build the database. Feedback is welcome.

Technically the main part happens [here](https://github.com/xemle/home-gallery/blob/feature/exif-tags/packages/database/src/map-media.js#L57). Limits are when the main file occurs multiple times in your media folder. E.g.

```
/folder/IMG_1234.jpg
/folder/IMG_1234.jpg.xmp
/other/folder/IMG_1234.jpg
```
Than the sidecar changes (e.g. by external media viewer) and incremental imports of HomeGallery can cause IPTC keywords out of sync. Single file occurrence, static/unchanged files or full imports should work as expected.

**Comment by smolin** [comment]
I finally got a change to test this out, and it looks great, thanks!

I did notice that this build, home-gallery-feature-exif-tags-latest-linux-x64, seems to behave differently from the last release I used --- this one extracts itself into /tmp/caxa/ and runs from there --- at least I think that's what it's doing. Maybe this is just done for testing but not in the main release builds ...

**Comment by xemle** [comment]
Thank you for testing. Does this exif meta data keyword feature covers your requirements? Do you miss something?

I am still playing around with it and take the change to migrate my old meta data. So thank you for requesting IPTC keywords ;-)

Regarding `/tmp/caxa` it is on purpose. The gallery uses the npm caxa package for bundling OS specific binaries. Node and all required files are extracted on the first run to that dir and the software is started from there.

**Comment by smolin** [comment]
I believe that will cover my requirements for now, thanks. I am thinking about running facial recognition locally, but I think that would end up in the same metadata keyword field.

Re caxa, interesting! I hadn't come across that package before, will read up on it.

Thanks again!

**Comment by xemle** [comment]
Hi @smolin version 1.4.1 is just released with XMP/IPTC keyword support. Please do not hesitate to test the new release. Any feedback is welcome.

**Comment by xemle** [comment]
... and the search capabilities have been improved as well. Searches by tags can be queried more explicitly. See [Search docs](https://docs.home-gallery.org/webapp/index.html#search) for details.

**Comment by smolin** [comment]
Thanks! I'll check it out

**Comment by xemle** [comment]
Close it since it was implemented 1.4.1

@smolin If you discover any problems, please do not hesitate to reopen this issue.
